design-proposals/generated-disk-serial.md

```diff
@@ -0,0 +1,202 @@
+- Generating the serial reduces manual configuration.
+
+## Goals
+- Add an option to the VM's `.spec` that will enable automatic serial number generation for disks that
+  don't have a serial number specified.
+
+- Generate unique serial numbers for each disk. The number has to be unique compared to all other disks attached
+  to all VMs with this disk.
+ 
+- Don't store the serial number in the VM's `.spec`. It would confuse gitops operators.
+
+- The serial number should remain the same across storage migration.
+
+## Non Goals
+
+- No customization options for serial number format.
+
+## Definition of Users
+
+- VM owners
+
+## User Stories
+
+- As a VM owner, I want to use the automated disk serial number assignment to reduce manual configuration.
+- As a VM owner, I want each disk to have a unique serial number.
+- As a VM owner, I want newly created Windows VMs to have serial numbers set on disks by default.
+
+## Repos
+- [kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+A new boolean option `generateDiskSerials` will be added to the VMI `.spec`. 
+When it is set to `true`, kubevirt will generate a serial number for each disk, that does not have one specified.
+These serial numbers will then be set in the libvirt domain xml.
+
+This option can also be added to `VirtualMachinePreference`, so it can be set on new VMs and
+the user does not need to be aware of it.
+
+## Persistence of generated serial numbers
+
+The serial number has to be the same across VM restarts and storage migration,
+so it has to be stored somewhere, or generated deterministically.
+
+This section lists several options how to store serial numbers.
+Then the rest of the document describes details of option: **Annotation on the PVC object**.
+
+### Annotation on the PVC object
+The serial number will be stored as an annotation on the PVC object. It would be created when the VMI is created.
+Then it will be read by `virt-handler` when creating libvirt domain XML.
+
+#### Advantages:
+- Serial is associated with PVC.
+- Serial is clearly visible to anyone who can read PVC.
+
+#### Disadvantages:
+- Does not store serials for other volume types except PVC.
+
+
+### Storing it in the `status` of the VM
+Serial numbers will be stored in the `.status` of the VM.
+GitOps does not use the `/status` subresource to update the objects it creates, so it should not overwrite the serials.
+More information about `.status` is [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status).
+
+#### Advantages:
+- We can store serial numbers for all volume types.
+- Serial is clearly visible to anyone who can read VM.
+
+#### Disadvantages:
+- Harder to make sure that a disk has the same serial if it is mounted to multiple VMs.
+- It may be an unexpected place to store persistent information that cannot be regenerated if `.status` is wiped.
+
+
+### Deterministic generation
```

Comment:
I like this approach the most. Basically a stateless approach without configurables.

Reply:
It would be the best choice, if it can be compatible with storage migration.